bokehjs/src/coffee/renderer/tile/mercator_tile_source.coffee

```diff
@@ -0,0 +1,241 @@
+    for key, tile of @tiles
+      if tile.quadkey.indexOf(quadkey) == 0 and tile.quadkey.length > min_zoom and tile.quadkey.length <= max_zoom
+        tile.retain = true
+
+  retain_neighbors:(reference_tile) ->
+    neighbor_radius = 4
+    [tx, ty, tz] = reference_tile.tile_coords
+    neighbor_x = (x for x in [tx - neighbor_radius .. tx + neighbor_radius])
+    neighbor_y = (y for y in [ty - neighbor_radius .. ty + neighbor_radius])
+
+    for key, tile of @tiles
+      if tile.tile_coords[2] == tz and _.contains(neighbor_x, tile.tile_coords[0]) and _.contains(neighbor_y, tile.tile_coords[1])
+        tile.retain = true
+
+  retain_parents:(reference_tile) ->
+    quadkey = reference_tile.quadkey
+    for key, tile of @tiles
+      tile.retain = quadkey.indexOf(tile.quadkey) == 0
+
+  children_by_tile_xyz: (x, y, z) ->
+    quad_key = @tile_xyz_to_quadkey(x, y, z)
+    child_tile_xyz = []
+    for i in [0..3] by 1
+      [x, y, z] = @quadkey_to_tile_xyz(quad_key + i.toString())
+      b = @get_tile_meter_bounds(x, y, z)
+      if b?
+        child_tile_xyz.push([x, y, z, b])
+    return child_tile_xyz
+
+  parent_by_tile_xyz: (x, y, z) ->
+    quad_key = @tile_xyz_to_quadkey(x, y, z)
+    parent_quad_key = quad_key.substring(0, quad_key.length - 1)
+    return @quadkey_to_tile_xyz(parent_quad_key)
+
+  get_resolution: (level) ->
+    return @get('initial_resolution') / Math.pow(2, level)
+
+  get_resolution_by_extent: (extent, height, width) ->
+    x_rs = (extent[2] - extent[0]) / width
+    y_rs = (extent[3] - extent[1]) / height
+    return [x_rs, y_rs]
+
+  get_level_by_extent: (extent, height, width) ->
+    x_rs = (extent[2] - extent[0]) / width
+    y_rs = (extent[3] - extent[1]) / height
+    resolution = Math.max(x_rs, y_rs)
+    i = 0
+    for r in @get('resolutions')
+      if resolution > r
+        return 0 if i == 0
+        return i - 1 if i > 0
+      i += 1
+
+  get_closest_level_by_extent:(extent, height, width) ->
+    x_rs = (extent[2] - extent[0]) / width
+    y_rs = (extent[3] - extent[1]) / height
+    resolution = Math.max(x_rs, y_rs)
+    ress = @get('resolutions')
+    closest = @get('resolutions').reduce (previous, current) ->
+      return current if (Math.abs(current - resolution) < Math.abs(previous - resolution))
+      return previous
+    return @get('resolutions').indexOf(closest)
+
+  snap_to_zoom: (extent, height, width, level) ->
+    desired_res = @get('resolutions')[level]
+    desired_x_delta = width * desired_res
+    desired_y_delta = height * desired_res
+
+    [xmin, ymin, xmax, ymax] = extent
+    x_adjust = (desired_x_delta - (xmax - xmin)) / 2
+    y_adjust = (desired_y_delta - (ymax - ymin)) / 2
+
+    return [xmin - x_adjust, ymin - y_adjust, xmax + x_adjust, ymax + y_adjust]
+
+  tms_to_wmts: (x, y, z) ->
+    '''Note this works both ways'''
+    return [x, 2 ** z - 1 - y, z]
+
+  wmts_to_tms: (x, y, z) ->
+    '''Note this works both ways'''
+    return [x, 2 ** z - 1 - y, z]
```

Comment:
This is simply a flipping of the Y-axis.  I included the function twice to help readability for logic where this is used.